tests/test_init.py

```diff
@@ -53,3 +60,99 @@ def test_parse_bad_patchwork_url(self):
             "Can't parse patchwork url: '{}'".format(patchwork_url)
             in context.exception
         )
+
+
+class KBuilderTest(unittest.TestCase):
+    # (Too many instance attributes) pylint: disable=R0902
+    """Test cases for skt.kbuilder class"""
+    def setUp(self):
+        self.tmpdir = tempfile.mkdtemp()
+        self.tmpconfig = tempfile.NamedTemporaryFile()
+        self.kbuilder = skt.kbuilder(self.tmpdir, self.tmpconfig.name)
+        self.m_popen = Mock()
+        self.m_popen.returncode = 0
+        self.ctx_popen = mock.patch('subprocess.Popen',
+                                    Mock(return_value=self.m_popen))
+        self.ctx_check_call = mock.patch('subprocess.check_call', Mock())
+        self.m_io_open = Mock()
+        self.m_io_open.__enter__ = lambda *args: self.m_io_open
+        self.m_io_open.__exit__ = lambda *args: None
+        self.ctx_io_open = mock.patch('io.open',
+                                      Mock(return_value=self.m_io_open))
+        self.kernel_tarball = 'linux-4.16.0.tar.gz'
+        self.success_str = 'Tarball successfully created in ./{}\n'
+        self.success_str = self.success_str.format(self.kernel_tarball)
+
+    def tearDown(self):
+        shutil.rmtree(self.tmpdir)
+
+    def test_mktgz_clean(self):
+        """
+        Check if `mrproper` target is called when `clean` is True and not
+        called when `clean` is False.
+        """
+        with self.ctx_popen, self.ctx_check_call as m_check_call:
+            self.assertRaises(Exception, self.kbuilder.mktgz, clean=True)
+            self.assertIn('make', m_check_call.mock_calls[0][1][0])
+            self.assertIn('mrproper', m_check_call.mock_calls[0][1][0])
+            m_check_call.reset_mock()
+            self.assertRaises(Exception, self.kbuilder.mktgz, clean=False)
+            self.assertNotIn('make', m_check_call.mock_calls[0][1][0])
```

Comment:
Strictly-speaking, this should check that `make` is not in the first argument only, not in any position. And the reverse above.

Reply:
I think it's **not** going to be better than the previous version in terms of the test fragility, after all, it's checking the exact arguments using hardcoded indexing. And the output when it fails is less intuitive. So I propose two things here, or we remove this test case or back to the previous version.

Reply:
Hmm, perhaps I misunderstood what's being done here, I'm not familiar with Python's `Mock`. I read this specific line (100) as it's checking that there was a command executed, which had no exact string `make` as any element in the `argv` array. E.g. command "echo hello" would pass, command "make something" would fail, but command "echo make" would fail as well and the latter was my concern.

However, I see that this is not as trivial as I thought. Could you perhaps explain what you're doing here, and why you think changing it to what I suggest would be worse? Thank you.

Reply:
Yeah, np. In the beginning, you mentioned that the test case was fragile because it checked all the arguments and I got your point. But if you take a look of what we'll have now (after addressing your last comment) and what we had before, it doesn't seem to be better at least for me in the term of fragility.

I didn't push the following code yet. It retrieves the first call to subprocess.check_call with the anti-intuitive [0][1][0] indexing and then compares index [0] == 'make' and so on.
```python
            self.assertEqual('make', m_check_call.mock_calls[0][1][0][0])
            self.assertIn('mrproper', m_check_call.mock_calls[0][1][0][3])
..
            self.assertNotEqual(make', m_check_call.mock_calls[0][1][0][0])
            self.assertNotEqual('mrproper', m_check_call.mock_calls[0][1][0][3])
```
vs
```python
            self.assertEqual(
                m_check_call.mock_calls[0],
                mock.call(['make', '-C', self.tmpdir, 'mrproper'])
            )
...
            self.assertNotEqual(
                m_check_call.mock_calls[0],
                mock.call(['make', '-C', self.tmpdir, 'mrproper']),
            )
```
Both have hardcoded arguments and for me the previous one has a better output when it fails.
```
======================================================================
FAIL: test_mktgz_clean_2 (tests.test_init.KBuilderTest)
Check if `mrproper` target is called when clean is True
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/test_init.py", line 109, in test_mktgz_clean_2
    mock.call(['mke', '-C', self.tmpdir, 'mrproper'])
AssertionError: call(['make', '-C', '/tmp/tmpQ1LiQC', 'mrproper']) != call(['mke', '-C', '/tmp/tmpQ1LiQC', 'mrproper'])
```
vs
```
======================================================================
FAIL: test_mktgz_clean (tests.test_init.KBuilderTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "tests/test_init.py", line 96, in test_mktgz_clean
    self.assertEqual('mke', m_check_call.mock_calls[0][1][0][0])
AssertionError: 'mke' != 'make'
```
I don't think it is worse, but I am not convinced it is better either hehe. I'm used to seeing comparisons using mock.call, maybe that's why it sounds more intuitive to me. Anyway I want to see this PR merged soon so I'll take the way you prefer and if in the future we find a better way to solve this, we can change it :):

Reply:
Ah, I see, thank you, Agustin. OK, minor correction to what I actually
suggested about `mrproper`: it should be searched for among all arguments. So,
it should rather be:

    self.assertEqual('make', m_check_call.mock_calls[0][1][0][0])
    self.assertIn('mrproper', m_check_call.mock_calls[0][1][0])

However, I agree that readability of both assert and the failure output is bad
in this case, so I think it's OK for us to match the exact command line, and
change this test from time to time as we update it, if that means we can
actually understand the test code and its failures.

Take your pick on how you want this done :)

Reply:
Thank you Nikolai! I really like your reviews :smile:. Please take a look at it again to see if it is what you expected.